File: python_ibkr/safe_spread_rescue.py
"""Rescue SAFE local quand le strike choisi est rejeté pour spread."""

from __future__ import annotations

TWO_PHASE_PUT_WINDOW = 10
SAFE_SPREAD_RESCUE_STRIKES_ABOVE = 2
SAFE_SPREAD_RESCUE_STRIKES_BELOW = 4
SAFE_SPREAD_REJECT_PCT = 35.0
SAFE_SPREAD_ACCEPTABLE_PCT = 20.0
SAFE_MIN_YIELD_PCT = 0.5


def normalize_spread_pct_percent(raw: float | None) -> float | None:
    if raw is None:
        return None
    try:
        x = float(raw)
    except (TypeError, ValueError):
        return None
    if x >= 0 and x <= 1.0001:
        return x * 100.0
    return x


def premium_yield_pct(put: dict) -> float | None:
    strike = put.get("strike")
    prime = put.get("primeUsed")
    if strike is None or prime is None:
        return None
    try:
        strike_f = float(strike)
        prime_f = float(prime)
    except (TypeError, ValueError):
        return None
    if strike_f <= 0 or prime_f <= 0:
        return None
    py = put.get("premiumYield")
    if py is not None:
        try:
            py_f = float(py)
            if py_f > 0:
                return py_f * 100.0 if py_f <= 1.0 else py_f
        except (TypeError, ValueError):
            pass
    return (prime_f / strike_f) * 100.0


def is_safe_rejected_for_spread(spread_pct_raw: float | None) -> bool:
    spread = normalize_spread_pct_percent(spread_pct_raw)
    if spread is None:
        return False
    return spread > SAFE_SPREAD_REJECT_PCT


def is_spread_acceptable(spread_pct_raw: float | None) -> bool:
    spread = normalize_spread_pct_percent(spread_pct_raw)
    if spread is None:
        return False
    return spread <= SAFE_SPREAD_ACCEPTABLE_PCT


def build_safe_rescue_strike_window(
    original_strike: float,
    sorted_strikes: list[float],
    strikes_above: int = SAFE_SPREAD_RESCUE_STRIKES_ABOVE,
    strikes_below: int = SAFE_SPREAD_RESCUE_STRIKES_BELOW,
) -> list[float]:
    ladder = sorted({float(s) for s in sorted_strikes if s is not None})
    if not ladder or original_strike is None:
        return []
    strike = float(original_strike)
    try:
        idx = ladder.index(strike)
    except ValueError:
        idx = next((i for i, s in enumerate(ladder) if s >= strike), len(ladder))
        if idx > 0 and abs(ladder[idx - 1] - strike) < abs(ladder[idx] - strike if idx < len(ladder) else strike + 999):
            idx -= 1
    below = ladder[max(0, idx - strikes_below) : idx]
    above = ladder[idx + 1 : idx + 1 + strikes_above]
    return [s for s in (*below, *above) if s != strike]


def _rank_key(put: dict, ctx: dict) -> tuple:
    spread = normalize_spread_pct_percent(put.get("spreadPct"))
    yield_pct = premium_yield_pct(put)
    strike = float(put.get("strike") or 0)
    original = float(ctx.get("original_strike") or 0)
    spot = ctx.get("spot")
    spread_tier = 0 if is_spread_acceptable(spread) else (1 if spread is not None and spread <= 35 else 2)
    spread_val = spread if spread is not None else 999.0
    yield_ok = 0 if (yield_pct is not None and yield_pct >= SAFE_MIN_YIELD_PCT) else 1
    bid_ok = 0 if (put.get("bid") is not None and float(put.get("bid") or 0) > 0) else 1
    dist = 0.0
    if spot is not None and spot > 0:
        dist = -((float(spot) - strike) / float(spot)) * 100.0
    proximity = abs(strike - original) if original else 999.0
    return (spread_tier, spread_val, yield_ok, bid_ok, -dist, proximity, -strike)


def _is_safe_distance_ok(put: dict, ctx: dict) -> bool:
    strike = put.get("strike")
    lb = ctx.get("lower_bound")
    agg = ctx.get("aggressive_strike")
    spot = ctx.get("spot")
    if put.get("isBelowLowerBound") is False:
        return False
    if strike is not None and lb is not None and float(strike) >= float(lb):
        return False
    if strike is not None and agg is not None and float(strike) >= float(agg):
        return False
    if strike is not None and spot is not None and float(spot) > 0:
        dist_pct = ((float(spot) - float(strike)) / float(spot)) * 100.0
        if dist_pct < 2:
            return False
    return True


def evaluate_safe_rescue_candidates(
    *,
    original_safe: dict,
    put_by_strike: dict[float, dict],
    rescue_strikes: list[float],
    lower_bound: float | None,
    aggressive_strike: float | None,
    target_premium: float | None,
    spot: float | None,
) -> dict:
    original_spread = normalize_spread_pct_percent(original_safe.get("spreadPct"))
    ctx = {
        "original_strike": original_safe.get("strike"),
        "spot": spot,
        "lower_bound": lower_bound,
        "aggressive_strike": aggressive_strike,
    }
    checked: list[float] = []
    eligible: list[tuple[tuple, dict]] = []
    for strike in rescue_strikes:
        checked.append(strike)
        put = put_by_strike.get(float(strike))
        if put is None:
            continue
        spread = normalize_spread_pct_percent(put.get("spreadPct"))
        if spread is None or original_spread is None or not (spread < original_spread):
            continue
        if put.get("bid") is None or float(put.get("bid") or 0) <= 0:
            continue
        tgt = float(target_premium or 0)
        prime = float(put.get("primeUsed") or 0)
        if tgt > 0 and prime < tgt:
            continue
        yld = premium_yield_pct(put)
        if yld is None or yld < SAFE_MIN_YIELD_PCT:
            continue
        if not _is_safe_distance_ok(put, ctx):
            continue
        eligible.append((_rank_key(put, ctx), put))
    eligible.sort(key=lambda row: row[0])
    winner = eligible[0][1] if eligible else None
    return {"checked": checked, "winner": winner}


def attempt_safe_spread_rescue(
    *,
    safe_pc: dict | None,
    aggressive_pc: dict | None,
    put_data: list[dict],
    strikes: list[float],
    lower_bound: float | None,
    target_premium: float | None,
    spot: float | None,
) -> tuple[dict | None, dict]:
    empty = {
        "safeSpreadRescueTriggered": False,
        "safeOriginalStrike": None,
        "safeOriginalBidAsk": None,
        "safeOriginalSpread": None,
        "safeRescueCandidatesChecked": [],
        "safeRescueStrike": None,
        "safeRescueBidAsk": None,
        "safeRescueSpread": None,
        "safeRescueReason": None,
    }
    if safe_pc is None:
        return safe_pc, empty

    original_spread = normalize_spread_pct_percent(safe_pc.get("spreadPct"))
    empty.update(
        {
            "safeOriginalStrike": safe_pc.get("strike"),
            "safeOriginalBidAsk": f"{safe_pc.get('bid')}/{safe_pc.get('ask')}",
            "safeOriginalSpread": original_spread,
        }
    )

    if not is_safe_rejected_for_spread(safe_pc.get("spreadPct")):
        empty["safeRescueReason"] = "safe_not_rejected_for_spread"
        return safe_pc, empty

    put_by_strike = {float(p["strike"]): p for p in put_data if p.get("strike") is not None}
    rescue_strikes = build_safe_rescue_strike_window(
        float(safe_pc["strike"]),
        [float(s) for s in strikes if s is not None],
    )
    agg_strike = aggressive_pc.get("strike") if aggressive_pc else None
    evaluation = evaluate_safe_rescue_candidates(
        original_safe=safe_pc,
        put_by_strike=put_by_strike,
        rescue_strikes=rescue_strikes,
        lower_bound=lower_bound,
        aggressive_strike=agg_strike,
        target_premium=target_premium,
        spot=spot,
    )

    diag = {
        **empty,
        "safeSpreadRescueTriggered": True,
        "safeRescueCandidatesChecked": evaluation["checked"],
        "safeRescueReason": "no_acceptable_rescue_candidate",
    }

    winner = evaluation.get("winner")
    if winner is None:
        return safe_pc, diag

    rescued_spread = normalize_spread_pct_percent(winner.get("spreadPct"))
    diag.update(
        {
            "safeRescueStrike": winner.get("strike"),
            "safeRescueBidAsk": f"{winner.get('bid')}/{winner.get('ask')}",
            "safeRescueSpread": rescued_spread,
            "safeRescueReason": (
                "replaced_with_acceptable_spread"
                if is_spread_acceptable(winner.get("spreadPct"))
                else "replaced_with_cleaner_spread"
            ),
        }
    )

    rescued = dict(safe_pc)
    rescued.update(winner)
    rescued["status"] = "safe_selected"
    rescued["reason"] = "safe_spread_rescue_local"
    rescued["selectionReason"] = "safe_spread_rescue_local"
    rescued["safeSpreadRescueApplied"] = True
    rescued["safeSpreadRescueDiagnostics"] = diag
    return rescued, diag

File: app/calculations/safeSpreadRescue.js
/**
 * SAFE spread rescue — recherche locale autour du strike SAFE original
 * quand celui-ci est rejeté pour spread (gradeLeg REJECT).
 *
 * Fenêtre IBKR two-phase par défaut : 10 strikes sous l'agressif (voir TWO_PHASE_PUT_WINDOW).
 * Ce rescue n'élargit pas la fenêtre globale ; il regarde seulement ±2/+4 strikes
 * autour du SAFE choisi, en réutilisant putCandidates ou quotes déjà fetchées.
 */

/** Miroir Python TWO_PHASE_DEFAULT_PUT_WINDOW / IBKR_TWO_PHASE_PUT_WINDOW. */
export const TWO_PHASE_PUT_WINDOW = 10;

export const SAFE_SPREAD_RESCUE_STRIKES_ABOVE = 2;
export const SAFE_SPREAD_RESCUE_STRIKES_BELOW = 4;

/** Seuil gradeLeg REJECT (spread > 35 %). */
export const SAFE_SPREAD_REJECT_PCT = 35;

/** Spread « acceptable » pour le tri (grade B). */
export const SAFE_SPREAD_ACCEPTABLE_PCT = 20;

/** Rendement minimum SAFE (0,5 % par expiration). */
export const SAFE_MIN_YIELD_PCT = 0.5;

export function normalizeSpreadPctPercent(raw) {
  const x = Number(raw);
  if (!Number.isFinite(x)) return null;
  if (x >= 0 && x <= 1.0001) return x * 100;
  return x;
}

export function premiumYieldPctFromPut(put) {
  const strike = Number(put?.strike);
  const prime =
    Number(put?.primeUsed ?? put?.premiumUsed ?? put?.bid ?? put?.conservativePremium);
  if (!Number.isFinite(strike) || strike <= 0 || !Number.isFinite(prime) || prime <= 0) {
    return null;
  }
  const fromField = Number(put?.premiumYield ?? put?.weeklyYield);
  if (Number.isFinite(fromField) && fromField > 0) {
    return fromField <= 1 ? fromField * 100 : fromField;
  }
  return (prime / strike) * 100;
}

export function isSafeRejectedForSpread(spreadPctRaw) {
  const spreadPct = normalizeSpreadPctPercent(spreadPctRaw);
  if (spreadPct == null) return false;
  return spreadPct > SAFE_SPREAD_REJECT_PCT;
}

export function isSpreadAcceptable(spreadPctRaw) {
  const spreadPct = normalizeSpreadPctPercent(spreadPctRaw);
  if (spreadPct == null) return false;
  return spreadPct <= SAFE_SPREAD_ACCEPTABLE_PCT;
}

/**
 * Strikes de rescue : jusqu'à 2 au-dessus et 4 en-dessous sur l'échelle triée.
 */
export function buildSafeRescueStrikeWindow(
  originalStrike,
  sortedStrikes,
  strikesAbove = SAFE_SPREAD_RESCUE_STRIKES_ABOVE,
  strikesBelow = SAFE_SPREAD_RESCUE_STRIKES_BELOW
) {
  const strike = Number(originalStrike);
  if (!Number.isFinite(strike)) return [];
  const ladder = [...new Set((sortedStrikes || []).map(Number).filter(Number.isFinite))].sort(
    (a, b) => a - b
  );
  if (!ladder.length) return [];

  let idx = ladder.indexOf(strike);
  if (idx === -1) {
    idx = ladder.findIndex((s) => s >= strike);
    if (idx === -1) idx = ladder.length;
    else if (idx > 0 && Math.abs(ladder[idx - 1] - strike) < Math.abs(ladder[idx] - strike)) {
      idx -= 1;
    }
  }

  const below = ladder.slice(Math.max(0, idx - strikesBelow), idx);
  const above = ladder.slice(idx + 1, idx + 1 + strikesAbove);
  return [...below, ...above].filter((s) => s !== strike);
}

function putSpreadPct(put) {
  return normalizeSpreadPctPercent(put?.spreadPct ?? put?.liquidity?.spreadPct);
}

function isUsableBid(put) {
  const bid = Number(put?.bid ?? put?.primeUsed ?? put?.premiumUsed);
  return Number.isFinite(bid) && bid > 0;
}

function meetsMinPremium(put, targetPremium) {
  const prime = Number(put?.primeUsed ?? put?.premiumUsed ?? put?.bid ?? put?.conservativePremium);
  const tgt = Number(targetPremium);
  if (!Number.isFinite(prime) || !Number.isFinite(tgt) || tgt <= 0) return prime > 0;
  return prime >= tgt;
}

function isSafeDistanceOk(put, { lowerBound, aggressiveStrike, spot }) {
  const strike = Number(put?.strike);
  const lb = Number(lowerBound);
  const agg = Number(aggressiveStrike);
  const px = Number(spot);

  if (Number.isFinite(lb) && Number.isFinite(strike) && strike >= lb) return false;
  if (put?.isBelowLowerBound === false) return false;
  if (Number.isFinite(agg) && Number.isFinite(strike) && strike >= agg) return false;

  if (Number.isFinite(px) && px > 0 && Number.isFinite(strike)) {
    const distPct = ((px - strike) / px) * 100;
    if (distPct < 2) return false;
  }
  return true;
}

/**
 * Score de tri — plus petit = meilleur candidat.
 */
export function rankSafeRescueCandidate(put, ctx) {
  const spreadPct = putSpreadPct(put);
  const yieldPct = premiumYieldPctFromPut(put);
  const strike = Number(put?.strike);
  const originalStrike = Number(ctx.originalStrike);
  const spreadAcceptable = isSpreadAcceptable(spreadPct);
  const originalSpread = normalizeSpreadPctPercent(ctx.originalSpreadPct);

  const spreadTier = spreadAcceptable ? 0 : spreadPct != null && spreadPct <= 35 ? 1 : 2;
  const spreadValue = spreadPct ?? 999;
  const yieldOk = yieldPct != null && yieldPct >= SAFE_MIN_YIELD_PCT ? 0 : 1;
  const yieldGap =
    yieldPct != null && Number.isFinite(Number(ctx.targetPremium)) && Number(ctx.targetPremium) > 0
      ? Math.abs(yieldPct - SAFE_MIN_YIELD_PCT)
      : yieldOk ? 10 : 0;
  const bidOk = isUsableBid(put) ? 0 : 1;
  const dist =
    Number.isFinite(Number(ctx.spot)) && Number.isFinite(strike) && Number(ctx.spot) > 0
      ? -((Number(ctx.spot) - strike) / Number(ctx.spot)) * 100
      : strike;
  const proximity = Number.isFinite(originalStrike)
    ? Math.abs(strike - originalStrike)
    : 999;

  return [
    spreadTier,
    spreadValue,
    yieldOk,
    yieldGap,
    bidOk,
    -dist,
    proximity,
    -strike,
  ];
}

export function evaluateSafeRescueCandidates({
  originalSafeStrike,
  originalSpreadPct,
  putCandidates = [],
  rescueStrikeWindow = [],
  lowerBound = null,
  aggressiveStrike = null,
  targetPremium = null,
  spot = null,
}) {
  const originalStrike = Number(originalSafeStrike?.strike ?? originalSafeStrike);
  const originalSpread = normalizeSpreadPctPercent(
    originalSpreadPct ?? originalSafeStrike?.spreadPct ?? originalSafeStrike?.liquidity?.spreadPct
  );

  const byStrike = new Map();
  for (const put of putCandidates || []) {
    const k = Number(put?.strike);
    if (Number.isFinite(k)) byStrike.set(k, put);
  }

  const windowStrikes =
    rescueStrikeWindow.length > 0
      ? rescueStrikeWindow
      : buildSafeRescueStrikeWindow(originalStrike, [...byStrike.keys()]);

  const ctx = {
    originalStrike,
    originalSpreadPct: originalSpread,
    targetPremium,
    spot,
    lowerBound,
    aggressiveStrike: Number(aggressiveStrike?.strike ?? aggressiveStrike),
  };

  const checked = [];
  const eligible = [];

  for (const strike of windowStrikes) {
    const put = byStrike.get(Number(strike));
    checked.push(strike);
    if (!put) continue;

    const spreadPct = putSpreadPct(put);
    if (spreadPct == null || originalSpread == null || !(spreadPct < originalSpread)) continue;
    if (!isUsableBid(put)) continue;
    if (!meetsMinPremium(put, targetPremium)) continue;
    const yieldPct = premiumYieldPctFromPut(put);
    if (yieldPct == null || yieldPct < SAFE_MIN_YIELD_PCT) continue;
    if (!isSafeDistanceOk(put, ctx)) continue;

    eligible.push({ put, spreadPct, yieldPct, rank: rankSafeRescueCandidate(put, ctx) });
  }

  eligible.sort((a, b) => {
    for (let i = 0; i < a.rank.length; i++) {
      if (a.rank[i] !== b.rank[i]) return a.rank[i] - b.rank[i];
    }
    return 0;
  });

  return { checked, eligible, winner: eligible[0] ?? null };
}

function formatBidAsk(put) {
  const bid = put?.bid ?? null;
  const ask = put?.ask ?? null;
  if (bid == null && ask == null) return null;
  return `${bid ?? "—"}/${ask ?? "—"}`;
}

/**
 * Tente un rescue SAFE local. Retourne le SAFE (original ou remplacé) + diagnostics.
 */
export function attemptSafeSpreadRescue({
  safeStrike,
  aggressiveStrike = null,
  putCandidates = [],
  lowerBound = null,
  targetPremium = null,
  spot = null,
  allStrikes = null,
  skipIfAlreadyRescued = true,
}) {
  const emptyDiagnostics = {
    safeSpreadRescueTriggered: false,
    safeOriginalStrike: null,
    safeOriginalBidAsk: null,
    safeOriginalSpread: null,
    safeRescueCandidatesChecked: [],
    safeRescueStrike: null,
    safeRescueBidAsk: null,
    safeRescueSpread: null,
    safeRescueReason: null,
  };

  if (!safeStrike || typeof safeStrike !== "object") {
    return { safeStrike, diagnostics: emptyDiagnostics };
  }

  if (skipIfAlreadyRescued && safeStrike?.safeSpreadRescueApplied === true) {
    return {
      safeStrike,
      diagnostics: safeStrike.safeSpreadRescueDiagnostics ?? {
        ...emptyDiagnostics,
        safeSpreadRescueTriggered: true,
        safeRescueReason: "already_applied",
      },
    };
  }

  const originalSpread = putSpreadPct(safeStrike);

  if (!isSafeRejectedForSpread(originalSpread)) {
    return {
      safeStrike,
      diagnostics: {
        ...emptyDiagnostics,
        safeOriginalStrike: safeStrike.strike ?? null,
        safeOriginalBidAsk: formatBidAsk(safeStrike),
        safeOriginalSpread: originalSpread,
        safeRescueReason: "safe_not_rejected_for_spread",
      },
    };
  }

  const ladder =
    allStrikes ??
    [...new Set((putCandidates || []).map((p) => Number(p?.strike)).filter(Number.isFinite))].sort(
      (a, b) => a - b
    );

  const rescueWindow = buildSafeRescueStrikeWindow(safeStrike.strike, ladder);
  const evaluation = evaluateSafeRescueCandidates({
    originalSafeStrike: safeStrike,
    originalSpreadPct: originalSpread,
    putCandidates,
    rescueStrikeWindow: rescueWindow,
    lowerBound,
    aggressiveStrike,
    targetPremium,
    spot,
  });

  const diagnostics = {
    safeSpreadRescueTriggered: true,
    safeOriginalStrike: safeStrike.strike ?? null,
    safeOriginalBidAsk: formatBidAsk(safeStrike),
    safeOriginalSpread: originalSpread,
    safeRescueCandidatesChecked: evaluation.checked,
    safeRescueStrike: null,
    safeRescueBidAsk: null,
    safeRescueSpread: null,
    safeRescueReason: evaluation.winner ? null : "no_acceptable_rescue_candidate",
  };

  if (!evaluation.winner) {
    return { safeStrike, diagnostics };
  }

  const winnerPut = evaluation.winner.put;
  const rescuedSpread = evaluation.winner.spreadPct;

  diagnostics.safeRescueStrike = winnerPut.strike ?? null;
  diagnostics.safeRescueBidAsk = formatBidAsk(winnerPut);
  diagnostics.safeRescueSpread = rescuedSpread;
  diagnostics.safeRescueReason = isSpreadAcceptable(rescuedSpread)
    ? "replaced_with_acceptable_spread"
    : "replaced_with_cleaner_spread";

  const rescuedSafe = {
    ...safeStrike,
    ...winnerPut,
    strike: winnerPut.strike,
    bid: winnerPut.bid ?? safeStrike.bid,
    ask: winnerPut.ask ?? safeStrike.ask,
    mid: winnerPut.mid ?? safeStrike.mid,
    spreadPct: winnerPut.spreadPct ?? safeStrike.spreadPct,
    primeUsed: winnerPut.primeUsed ?? winnerPut.bid ?? safeStrike.primeUsed,
    premiumUsed: winnerPut.primeUsed ?? winnerPut.premiumUsed ?? winnerPut.bid ?? safeStrike.premiumUsed,
    selectionReason: "safe_spread_rescue_local",
    safeSpreadRescueApplied: true,
    safeSpreadRescueDiagnostics: diagnostics,
  };

  return { safeStrike: rescuedSafe, diagnostics };
}

/**
 * Spread actif pour le rejet global « Rejetés pour spread » selon le mode sélectionné.
 */
export function getActiveSpreadPctForSelectedMode({
  safeSpreadPct = null,
  aggressiveSpreadPct = null,
  selectedMode = null,
  card = null,
}) {
  const safe = normalizeSpreadPctPercent(
    safeSpreadPct ??
      card?.safeStrike?.liquidity?.spreadPct ??
      card?.safeStrike?.spreadPct ??
      card?.safeSpreadPct
  );
  const agg = normalizeSpreadPctPercent(
    aggressiveSpreadPct ??
      card?.aggressiveStrike?.liquidity?.spreadPct ??
      card?.aggressiveStrike?.spreadPct
  );

  let mode = String(selectedMode || "").trim().toUpperCase();
  if (!mode && card) {
    mode = String(
      card.finalDisplayMode ?? card.selectedMode ?? card.recommendedMode ?? ""
    ).toUpperCase();
  }

  if (mode === "AGGRESSIVE") return agg;
  if (mode === "SAFE") return safe;

  if (mode === "BALANCED" && card) {
    const safeLeg = card.safeStrike ?? card._safeLeg ?? null;
    const aggLeg = card.aggressiveStrike ?? card._aggLeg ?? card.selectedLeg ?? null;
    const safeY = Number(safeLeg?.weeklyYield ?? card._safeYieldPct);
    const aggY = Number(aggLeg?.weeklyYield ?? card._aggYieldPct);
    const hasSafe = safeLeg != null;
    const hasAgg = aggLeg != null;
    const safeInRange = hasSafe && Number.isFinite(safeY) && safeY >= 0.75 && safeY < 1.05;
    const aggInRange = hasAgg && Number.isFinite(aggY) && aggY >= 0.75 && aggY < 1.05;
    const mid = 0.875;
    if (safeInRange && aggInRange) {
      return Math.abs(safeY - mid) <= Math.abs(aggY - mid) ? safe : agg;
    }
    if (safeInRange) return safe;
    if (aggInRange) return agg;
    if (hasAgg) return agg;
    if (hasSafe) return safe;
  }

  if (card?.finalDisplayMode === "AGGRESSIVE") return agg;
  if (card?.finalDisplayMode === "SAFE") return safe;

  const values = [safe, agg].filter((v) => Number.isFinite(v));
  return values.length ? Math.max(...values) : null;
}

export function shouldGlobalSpreadReject(card, { extremeThresholdPct = 80 } = {}) {
  const safe = normalizeSpreadPctPercent(
    card?.safeStrike?.liquidity?.spreadPct ??
      card?.safeStrike?.spreadPct ??
      card?.safeSpreadPct
  );
  const agg = normalizeSpreadPctPercent(
    card?.aggressiveStrike?.liquidity?.spreadPct ?? card?.aggressiveStrike?.spreadPct
  );
  const active = getActiveSpreadPctForSelectedMode({
    safeSpreadPct: safe,
    aggressiveSpreadPct: agg,
    card,
  });
  return active != null && active > extremeThresholdPct;
}
